src/Compilers/Server/VBCSCompiler/NamedPipeClientConnection.cs

```diff
@@ -5,270 +5,84 @@
-                }
-                else if (!allowCompilationRequests)
-                {
-                    return await HandleRejectedRequestAsync("Compilation requests not allowed at this time", cancellationToken).ConfigureAwait(false);
-                }
-                else
-                {
-                    return await HandleCompilationRequestAsync(request, cancellationToken).ConfigureAwait(false);
-                }
-            }
-            finally
-            {
-                Close();
+                IsDisposed = true;
             }
         }
 
-        private async Task<ConnectionData> HandleCompilationRequestAsync(BuildRequest request, CancellationToken cancellationToken)
+        public async Task<BuildRequest> ReadBuildRequestAsync(CancellationToken cancellationToken)
         {
-            var keepAlive = CheckForNewKeepAlive(request);
-
-            // Kick off both the compilation and a task to monitor the pipe for closing.
-            var buildCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-            var compilationTask = ServeBuildRequestAsync(request, buildCts.Token);
-            var monitorTask = MonitorDisconnectAsync(buildCts.Token);
-            await Task.WhenAny(compilationTask, monitorTask).ConfigureAwait(false);
+            var request = await BuildRequest.ReadAsync(Stream, cancellationToken).ConfigureAwait(false);
 
-            // Do an 'await' on the completed task, preference being compilation, to force
-            // any exceptions to be realized in this method for logging.
-            CompletionReason reason;
-            if (compilationTask.IsCompleted)
-            {
-                var response = await compilationTask.ConfigureAwait(false);
-
-                try
-                {
-                    await response.WriteAsync(_stream, cancellationToken).ConfigureAwait(false);
-                    reason = CompletionReason.CompilationCompleted;
-                }
-                catch
-                {
-                    reason = CompletionReason.ClientDisconnect;
-                }
-            }
-            else
+            // Now that we've read data from the stream we can validate the identity.
+            if (!NamedPipeUtil.CheckClientElevationMatches(Stream))
             {
-                await monitorTask.ConfigureAwait(false);
-                reason = CompletionReason.ClientDisconnect;
+                throw new Exception("Client identity does not match server identity.");
             }
 
-            // Begin the tear down of the Task which didn't complete.
-            buildCts.Cancel();
-            return new ConnectionData(reason, keepAlive);
-        }
+            // The result is deliberately discarded here. The idea is to kick off the monitor code and 
+            // when it completes it will trigger the task. Don't want to block on that here.
+            _ = MonitorDisconnect();
```

Comment:
Won't this mean that exceptions from this cannot ever be handled sanely? Or is the thought that no matter what, exceptions from here should just bring down the server?

Reply:
The underlying call here should never throw. That is part of the contract. That is a bit flaky though. Decided to wrap the underyling call in a `try / catch` to make the contract stronger.